### PR TITLE
Support local buildpacks and meta-buildpacks in libcnb-test

### PR DESCRIPTION
**Pull request by colincasey**
The test runner has been modified to support a `BuildpackReference::LibCnbRs(BuildpackId)` variant which represents a buildpack (and any of it's dependencies) within the workspace that needs to be compiled for testing. This is similar to `BuildpackReference::Crate` but also supports meta-buildpacks.

This PR replaces #590 and should be easier to review now that similar refactorings from that PR have been incorporated into the codebase.

### CODE REVIEW

File: libcnb-test/src/test_runner.rs
Comment (edmorley) [inline libcnb-test/src/test_runner.rs:59]: I think we should drop the Pack CLI version check, since:
- there's no way to only run it once before all tests (due to the way Cargo integration tests work), so we have to run it before every test
- our CI is already using Pack CLI 0.30 (thanks to https://github.com/buildpacks/github-actions already having updated the default version to it)
- this just leaves buildpack maintainers who will need to update - and realistically just the languages team at the moment (in the future libcnb-test might be used elsewhere, but by then this 0.30+ version check will be redundant and removed regardless)
- since the bug only affects this new feature (meta-buildpack support), we could just mention the pack 0.30+ requirement in the changelog entry for the new feature (+ also the rustdocs for `BuildpackReference::LibCnbRs`)

File: libcnb-test/src/build_config.rs
Comment (edmorley) [inline libcnb-test/src/build_config.rs:258]: I find the current names a bit confusing - is there a way we can improve these?

Crate -> CurrentCrate?
LibCnbRs -> LocalCrate?


...or something else?

File: .gitignore
Comment (edmorley) [inline .gitignore:8]: We have quite a few different style of naming for the fixture directories - we should fix this at some point later (in a separate PR)

File: libcnb-test/src/build_config.rs
Comment (edmorley) [inline libcnb-test/src/build_config.rs:252]: The new feature (plus enum variant) needs a changelog entry

File: libcnb-test/src/build_config.rs
Comment (colincasey) [reply libcnb-test/src/build_config.rs:258]: i'm open to suggestions. if we want to keep the language close to Rust-isms (e.g.; `Crate`, `WorkspaceMember`) that kind of makes sense. it is a little weird when those combine with Buildpack-isms though because not everything is a crate. 

For example:
* using `WorkspaceMember("heroku/nodejs-engine")` or `LocalCrate("heroku/nodejs-engine")` for a component buildpack where `heroku/nodejs-engine` is the Buildpack Id & `buildpacks/nodejs-engine` is the Workspace path.
* using `WorkspaceMember("heroku/nodejs")` or `LocalCrate("heroku/nodejs")` for a composite buildpack where `heroku/nodejs` is the Buildpack Id & `meta-buildpacks/nodejs` is the Workspace path but it isn't a crate.

to avoid that mismatch what do you think of `WorkspaceBuildpack(BuildpackId)` @edmorley?

File: .gitignore
Comment (colincasey) [reply .gitignore:8]: i think this could be fixed if the buildpacks for the tests were copied into a temp folder and then used but, for that to work completely, there would need to be some way of setting the workspace directory also.  right now it defaults to the `libcnb.rs` workspace because that is where the integration tests run.

Comment (edmorley) [reply .gitignore:8]: I think you might have misread my comment - that wasn't what I was referring to? :-)

File: libcnb-test/src/build_config.rs
Comment (edmorley) [reply libcnb-test/src/build_config.rs:258]: Oh true, of course the meta-buildpacks aren't a crate.

`WorkspaceBuildpack` sounds good to me.

What do you think about also renaming `Crate` to `CurrentCrate`?

File: .gitignore
Comment (colincasey) [reply .gitignore:8]: yeah, my reply was more around the context of why i needed to add this line (which i was hoping to avoid), not the inconsistencies with fixture directory naming 😅

File: libcnb-test/src/build_config.rs
Comment (colincasey) [reply libcnb-test/src/build_config.rs:252]: Added

**Review by edmorley — COMMENTED** [review]
Thank you for making those changes. I'll defer to Manuel for the main part of the review, since he's more familiar with the libcnb-package APIs than I am.

**Comment by Malax** [comment]
Update: This is on my short list for this week for a review. I should get to this tomorrow.

File: libcnb-test/src/build.rs
Comment (colincasey) [inline libcnb-test/src/build.rs:74]: @Malax curious about your thoughts here.

File: libcnb-test/src/build.rs
Comment (Malax) [inline libcnb-test/src/build.rs:125]: It would be good to not have to have this additional struct, I removed most of these grouping structs from `libcnb-package` because they're confusing. We can do so easily by not creating a temporary directory in `package_buildpack` but instead pass in a target directory explicitly. The responsibility to create the temporary directory would automatically "bubble up" and end up in the place where the context of a single test lives.

All in all, we gain:
- A better isolated `package_buildpack` that does less implicit things.
  - Which also makes it easier to test (if we want to) or use in other contexts.
- Removal of the struct and, more importantly, the congnitive burden of having to remember what `PackagedBuildpackDir` actually means. When I saw the name for the first time, I immediately jumped to "it can't be just a path, otherwise if would be a `PathBuf`". But the name doesn't give a hint what that extra bit is.


Note: This will also effect `package_crate_buildpack`.

File: libcnb-test/src/build.rs
Comment (Malax) [inline libcnb-test/src/build.rs:132]: `CannotReadBuildpack` what exactly? Given that it wraps a `TomlFileError` is probably the buildpack descriptor? We should use a more concrete name here I think.

File: libcnb-test/src/build.rs
Comment (Malax) [inline libcnb-test/src/build.rs:135]: When I see lines like this when I wrote them, I rename either one. If `FindCargoWorkspaceRoot` isn't a good name we should fix that instead of using a slightly different name for the (in this case) enum variant. If we think `FindCargoWorkspaceRoot` is fine, there is no reason why the enum variant should be named differnetly.

File: libcnb-test/src/build.rs
Comment (Malax) [inline libcnb-test/src/build.rs:133]: Same as below, if there is a good reason to name this "create" instead of "build" we should rename `BuildBuildpackDependencyGraphError`. If there is no reason, the enum variant should follow the existing naming.

File: libcnb-test/src/build.rs
Comment (Malax) [inline libcnb-test/src/build.rs:20]: What up with this rename from `PackageCrateBuildpackError` to `PackageTestBuildpackError`? The previous name matches the error name. With this renaming its no longer the case. Should we rename the function too?

File: libcnb-test/tests/integration_test.rs
Comment (Malax) [inline libcnb-test/tests/integration_test.rs:68]: Why are we not continuing to use `should_panic` here (and in the other tests)?

**Review by Malax — CHANGES_REQUESTED** [review]
Overall good to go I think - some minor changes requested but nothing crazy I think.

File: libcnb-test/src/build.rs
Comment (colincasey) [reply libcnb-test/src/build.rs:74]: Addressed in https://github.com/heroku/libcnb.rs/pull/673

File: libcnb-test/src/build.rs
Comment (colincasey) [reply libcnb-test/src/build.rs:20]: I consider `package_crate_buildpack` to be a specialized version of `package_buildpack`.  All the errors are the same between both functions. Also, not all buildpacks are crates (i.e.; meta-buildpacks) which is why I chose to rename to `PackageTestBuildpackError` and don't see the need for the error name to align with the function names here.

File: libcnb-test/src/build.rs
Comment (colincasey) [reply libcnb-test/src/build.rs:125]: 👍 Good point.  The only reason I introduced this wrapper struct was so that the temporary directory wouldn't immediately go out of scope which would cause the buildpacks to be cleaned up before the actual test executes.  But it's not needed if I just move the temp dir creation to the runner.

File: libcnb-test/src/build.rs
Comment (colincasey) [reply libcnb-test/src/build.rs:132]: renamed to `CannotReadBuildpackDescriptor`

File: libcnb-test/src/build.rs
Comment (colincasey) [reply libcnb-test/src/build.rs:133]: renamed to `BuildBuildpackDependencyGraph`

File: libcnb-test/src/build.rs
Comment (colincasey) [reply libcnb-test/src/build.rs:135]: Most of my error names were carried over from my initial PR for this which had slightly different names for these same functions before refactoring. Renaming to `FindCargoWorkspaceRoot`

File: libcnb-test/src/build.rs
Comment (colincasey) [reply libcnb-test/src/build.rs:20]: @Malax given the above, would you be okay with `PackageTestBuildpackError` or `PackageBuildpackError`?

File: libcnb-test/tests/integration_test.rs
Comment (colincasey) [reply libcnb-test/tests/integration_test.rs:68]: Because this specific error now happens later on in the packaging process.  The error that happens now that we're reusing packaging logic is:
```
Could not package directory as buildpack! No `buildpack.toml` file exists 
at /some/unknown/directory/where/this/test/runs
```

To capture that error I can't use the `should_panic` macro, right?

File: libcnb-test/README.md
Comment (edmorley) [inline libcnb-test/README.md:194]: (End users won't have libcnb_data in their direct dependencies list, but instead libcnb)

```suggestion
use libcnb::data::buildpack_id;
```

File: libcnb-test/tests/integration_test.rs
Comment (edmorley) [inline libcnb-test/tests/integration_test.rs:549]: Since the test buildpacks always pass detection and don't use the procfile, they can use the empty fixture (using the procfile fixture but not the procfile buildpack made me wonder if there was something special about the procfile that meant it was still needed).

```suggestion
        BuildConfig::new("heroku/builder:22", "test-fixtures/empty").buildpacks(vec![
```

File: libcnb-test/tests/integration_test.rs
Comment (edmorley) [inline libcnb-test/tests/integration_test.rs:568]:
```suggestion
        BuildConfig::new("heroku/builder:22", "test-fixtures/empty").buildpacks(vec![
```

File: libcnb-test/README.md
Comment (colincasey) [reply libcnb-test/README.md]: fyi - to make this change work i had to add `libcnb` as a dev dependency to `libcnb-test/Cargo.toml`

File: libcnb-test/src/build.rs
Comment (Malax) [reply libcnb-test/src/build.rs]: `PackageBuildpackError` seems good to me. Thanks for the explanation! :)

File: libcnb-test/tests/integration_test.rs
Comment (Malax) [reply libcnb-test/tests/integration_test.rs:68]: Hm, I'm not sure. This is not blocking for me, let's revisit this later after merging. :)

**Review by Malax — APPROVED** [review]
Approve to stop blocking, feel free to merge after fixing the two comments I made. :) 👍🏻

File: libcnb-test/Cargo.toml
Comment (colincasey) [reply libcnb-test/Cargo.toml]: newline added

File: libcnb-test/tests/integration_test.rs
Comment (edmorley) [reply libcnb-test/tests/integration_test.rs:68]: > Why are we not continuing to use `should_panic` here (and in the other tests)?

So it turns out the reason is that the top level error handling had regressed, which is why the test only passed when using the lower level error: #704.

**Pull request merged**
